dss/util/iterators.py

```diff
@@ -0,0 +1,177 @@
+from itertools import tee
+from typing import Iterator, TypeVar, Optional, Tuple, Iterable, NamedTuple
+
+T = TypeVar('T')
+OT = Optional[T]
+Column = Iterator[T]
+
+
+class zipalign(Iterator['zipalign.Row']):
```

Comment:
Echoing the need for unit tests of zipalign.

Also, just to throw my hat into the algo-refactor ring, I wonder if it might be simpler to just advance the minimum paging iterator:
```
import random

LARGER_THAN_ANYTHING = 1e9

def get_list():
    foo = set([random.randint(1, 20) for _ in range(10)])
    return sorted(list(foo))


lists = [get_list() for _ in range(3)]
missing = [list() for _ in range(len(lists))]
iters = [iter(l) for l in lists]

for l in lists:
    print(l)

curs = [next(i) for i in iters]
while True:
    m = min(curs)
    if not any(iters):
        break
    for i, val in enumerate([c for c in curs]):
        if val is None:
            continue

        if curs[i] > m:
            print(f"iter {i} missing {m}")
            missing[i].append(m)
        elif curs[i] == m:
            try:
                curs[i] = next(iters[i])
            except StopIteration:
                iters[i] = None
                curs[i] = LARGER_THAN_ANYTHING
                         
for i in range(len(lists)):
    lists[i].extend(missing[i])
    print(sorted(lists[i]))
```

output:
```
[1, 2, 4, 6, 7, 8, 11, 14, 16, 20]
[1, 4, 6, 8, 12, 14, 19, 20]
[3, 4, 9, 10, 12, 13, 14, 16, 17]
iter 2 missing 1
iter 1 missing 2
iter 2 missing 2
iter 0 missing 3
iter 1 missing 3
iter 2 missing 6
iter 1 missing 7
iter 2 missing 7
iter 2 missing 8
iter 0 missing 9
iter 1 missing 9
iter 0 missing 10
iter 1 missing 10
iter 1 missing 11
iter 2 missing 11
iter 0 missing 12
iter 0 missing 13
iter 1 missing 13
iter 1 missing 16
iter 0 missing 17
iter 1 missing 17
iter 0 missing 19
iter 2 missing 19
iter 2 missing 20
[1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 20]
[1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 20]
[1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 20]
```

Reply:
Guys, please, lets not waste precious developer time on each of us contributing a differently shaped wheel. My algorithm is linear in time and space both on number of keys and number of replicas. Prove that my solution is severely flawed or can be significantly simplified before spending more time on reimplementing partially it without significant gain.

Reply:
@xbrianh, as mentioned before, the doctests are the unit test. Coverage is 100%:

https://codecov.io/gh/HumanCellAtlas/data-store/compare/99d8b7c97aaa45376f3f0381290b3dbe5891009c...aeccc481ae966b96f244c071c812a258411f4106/diff#D9-9

Reply:
It's worth noting that I spent some time coming up with a suitable interface between the the general algorithm and the concrete use case. Without my efforts, there wouldn't be this abstraction that you all feel compelled to reinvent. If you want to come up with a functionally equivalent (!) and more readable or significantly more efficient solution, I urge you to do so in a subsequent PR after (!) justifying that effort by outlining your approach and showing its potential to the team.

Reply:
My example is probably algorithmically equivalent to other solutions presented here. Please read it as a clumsy comment about clarity and brevity, as I'm not sure I understand the need for external chunks of abstract logic in this case.